daceml/onnx/op_implementations/fpga_implementations.py

```diff
@@ -886,6 +886,1034 @@ def make_compute(sdfg, state, vec_width=1):
+        # A PE needs K*T cycles to compute the results and then P*T clock cycles
+        # to fully drain them (draining is distributed across PEs).
+        # Therefore, in order to guarantee correctness and deadlock free we have
+        # to ensure that the number of cycles needed to drain the results is less
+        # or equal to the number of cycles needed to compute them.
+        # That is PT <= KT.
+        if K is not None and P > K:
+            raise ValueError(
+                f"Conv Im2Col (tiled): Number of processing elements {P} must be smaller than the K-dimension {K}."
+            )
+
+        def make_read_A(state):
+            '''
+            A is the weights/features streamed to PEs for Im2Col layout i.e. *W*
+            '''
+
+            # A given row of A must be repeated according to B number of tiles
+            # Both N and M can be not a multiple of P and T respectively
+            entry, exit = state.add_map(
+                "read_A",
+                {
+                    "b": f"0:{batch_size}",  # additional batch map
+                    "n0": f"0:ceiling({N}/{P})",
+                    "tm": f"0:ceiling({M}/{T})",
+                    "k": f"0:{K}",
+                    "n1": f"0:{P}"
+                },
+                schedule=dace.ScheduleType.FPGA_Device)
+
+            # The reader of A reads one element per clock cycle.
+            # Note that if P > T+L, then this will be the bottleneck
+            mem = state.add_read("W")
+            pipe = state.add_write("A_pipe")
+
+            # Read data from memory: if we are out-of-bound do not read from memory
+            # but inject dummy data
+            tasklet = state.add_tasklet(
+                "read_A", {"from_memory"}, {"to_kernel"}, f"""\
+data = from_memory if n0 * {P} + n1 < {N} else 0
+to_kernel = data""")
+
+            # Access mapping for Im2Col
+            filter = f"n0 * {P} + n1"  # directly corresponds to rows of Im2Col matrix
+            in_channel = f"int_floor(k,({filter_hx} * {filter_hy}))"
+            hy = f"int_floor((k % ({filter_hx} * {filter_hy})), {filter_hx})"
+            hx = f"(k % ({filter_hx} * {filter_hy})) % {filter_hx}"
+
+            access = f"[{filter}, {in_channel}, {hy}, {hx}]"
+
+            state.add_memlet_path(mem,
+                                  entry,
+                                  tasklet,
+                                  dst_conn="from_memory",
+                                  memlet=dace.Memlet(f"W{access}",
+                                                     dynamic=True,
+                                                     allow_oob=True))
+            state.add_memlet_path(tasklet,
+                                  exit,
+                                  pipe,
+                                  src_conn="to_kernel",
+                                  memlet=dace.Memlet(f"A_pipe[{P} - n1 - 1]"))
+
+        def make_read_B(state):
```

Comment:
also here, I suggest to call it `make_read_X`

Reply:
addressed